### ISSUE TITLE
Still fails to compile on VS2013

### ISSUE THREAD

**Issue by ned14** [eggs-cpp/variant#3]
Using master branch I still see this compilation error (the pull request I sent did not fix this, I couldn't figure out a workaround without rewriting your visitor implementation):

```
2>G:\eggs.variant\include\eggs/variant/detail/visitor.hpp(45): error C2440: 'initializing' : cannot convert from 'overloaded-function' to 'const function_ptr'
2>          None of the functions with this name in scope match the target type
2>          G:\eggs.variant\include\eggs/variant/detail/visitor.hpp(319) : see reference to function template instantiation 'R eggs::variants::detail::visitor<eggs::variants::detail::_apply<R,F,eggs::variants::detail::pack<>,eggs::variants::detail::pack<_Ty &&>>,R (F,void *,V)>::operator ()<int&&,std::string&&>(eggs::variants::detail::pack<int &&,std::string &&>,size_t,fun2 &,void *&&,_Ty &&) const' being compiled
2>          with
2>          [
2>              R=size_t
2>  ,            F=fun2 &
2>  ,            _Ty=eggs::variants::variant<std::string,int>
2>  ,            V=eggs::variants::variant<std::string,int> &&
2>          ]
2>          G:\eggs.variant\include\eggs/variant/detail/visitor.hpp(320) : see reference to function template instantiation 'R eggs::variants::detail::visitor<eggs::variants::detail::_apply<R,F,eggs::variants::detail::pack<>,eggs::variants::detail::pack<_Ty &&>>,R (F,void *,V)>::operator ()<int&&,std::string&&>(eggs::variants::detail::pack<int &&,std::string &&>,size_t,fun2 &,void *&&,_Ty &&) const' being compiled
2>          with
2>          [
2>              R=size_t
2>  ,            F=fun2 &
2>  ,            _Ty=eggs::variants::variant<std::string,int>
2>  ,            V=eggs::variants::variant<std::string,int> &&
2>          ]
2>          G:\eggs.variant\include\eggs/variant/variant.hpp(1453) : see reference to function template instantiation 'R eggs::variants::detail::apply<R,fun2&,_Ty,eggs::variants::variant<std::string,int>>(F,V &&,eggs::variants::variant<std::string,int> &&)' being compiled
2>          with
2>          [
2>              R=size_t
2>  ,            _Ty=eggs::variants::variant<int,std::string>
2>  ,            F=fun2 &
2>  ,            V=eggs::variants::variant<int,std::string>
2>          ]
2>          ..\..\test\apply.cpp(240) : see reference to function template instantiation 'R eggs::variants::apply<size_t,fun2&,eggs::variants::variant<int,std::string>,eggs::variants::variant<std::string,int>,void>(F,eggs::variants::variant<int,std::string> &&,eggs::variants::variant<std::string,int> &&)' being compiled
2>          with
2>          [
2>              R=size_t
2>  ,            F=fun2 &
2>          ]
```

I can tell you that reducing the overload set e.g. having only const void * overloads fixes the problem. Or only void * overloads. Mixing the two gives the error.

This bug was also in VS2015 right up until the very most recent CTP release.

Niall

**Comment by K-ballo** [comment]
As I mentioned on #2, everything but the latest VS CTP still fail on the `apply` with rvalues test. I don't want to disable perfect forwarding for VS and have the semantics silently change, so I had considered completely disabling `apply` on VS. But that seems to drastic, non-rvalues work perfectly fine.

> I can tell you that reducing the overload set e.g. having only const void * overloads fixes the problem. Or only void * overloads. Mixing the two gives the error.

With this key piece of information, I was able to apply an ugly workaround by simply specializing the visitor on `const`-or-not. VS2013 takes it, I will commit once I test it on the CTP.

**Comment by K-ballo** [comment]
Fixed by e5d9fff171a4a6fc88c4c47175e732abf83b3ca1

**Comment by ned14** [comment]
FYI your "ugly workaround" is quite beautiful compared with my truly ugly hack here. I am glad that my solution was useful. And thanks for the fix, it's good to have mostly full VS2013 compat.
